src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -200,19 +201,67 @@ export class TerminalService implements ITerminalService {
 		this._connectionState = TerminalConnectionState.Connected;
 		// now that terminals have been restored,
 		// attach listeners to update remote when terminals are changed
-		this.attachRemoteListeners();
+		this.attachListeners(true);
 		this._onDidChangeConnectionState.fire();
 	}
 
-	private attachRemoteListeners(): void {
+	private _reconnectToLocalTerminals(): void {
+		// Reattach to all local terminals
+		const defaultLayoutInfo = JSON.stringify({ tabs: [{ terminals: [], isActive: true }] });
+		const result = this._storageService.get('localTerminalLayoutInfo', StorageScope.WORKSPACE, defaultLayoutInfo);
+		const layoutInfo = JSON.parse(result);
+
+		if (layoutInfo.tabs.length === 0) {
+			this.createTerminal();
+			this.attachListeners(false);
+			return;
+		}
+
+		let activeTab: ITerminalTab | undefined;
+		if (layoutInfo) {
+			for (const layout of layoutInfo.tabs) {
+				let terminalInstance: ITerminalInstance | undefined;
+				let tab: ITerminalTab | undefined;
+				for (let i = 0; i < layout.terminals.length; i++) {
+					if (!terminalInstance) {
+						// create tab and terminal
+						terminalInstance = this.createTerminal();
+						tab = this._getTabForInstance(terminalInstance);
+						if (layout.isActive) {
+							activeTab = tab;
+						}
+					} else {
+						// add split terminal to this tab
+						this.splitInstance(terminalInstance);
+					}
+				}
+				// TODO@meganrogge: use processId instead of id eventually
+				const activeInstance = this.terminalInstances.find(t => t.id === layout.activeTerminalProcessId);
+				if (activeInstance) {
+					this.setActiveInstance(activeInstance);
+				}
+				tab?.resizePanes(layout.terminals.map((terminal: { relativeSize: number; }) => terminal.relativeSize));
+			}
+			if (layoutInfo.tabs.length) {
+				this.setActiveTabByIndex(activeTab ? this.terminalTabs.indexOf(activeTab) : 0);
+			}
+		}
+		this._connectionState = TerminalConnectionState.Connected;
+		// now that terminals have been restored,
+		// attach listeners to update local state when terminals are changed
+		this.attachListeners(false);
+		this._onDidChangeConnectionState.fire();
+	}
+
+	private attachListeners(isRemote: boolean): void {
 		this.onActiveTabChanged(() => {
-			this._updateRemoteState();
+			isRemote ? this._updateRemoteState() : this._updateLocalState();
 		});
 		this.onActiveInstanceChanged(() => {
-			this._updateRemoteState();
+			isRemote ? this._updateRemoteState() : this._updateLocalState();
 		});
 		this.onInstancesChanged(() => {
-			this._updateRemoteState();
+			isRemote ? this._updateRemoteState() : this._updateLocalState();
 		});
 	}
```

Comment:
Let's make this name a little more explicit, we can save some conditionals by moving the isRemote check higher too:

```ts
private attachProcessLayoutListeners(isRemote: boolean): void {
  if (isRemote) {
    this.onActiveTabChanged(() => this._updateRemoteState());
    this.onActiveInstanceChanged(() => this._updateRemoteState());
    this.onInstancesChanged(() => this._updateRemoteState());
  } else {
    ...
  }
}
```